### ISSUE TITLE
Missing field error in initializer for struct with inaccessible fields

### ISSUE THREAD

**Issue by gheoan** [rust-lang/rust#87872]
Given the following code:

```rust
// in file a.rs
#![crate_type = "lib"]

pub struct X {  
  pub f: u32, // remove the field f to get a different error message
  pub(crate) g: u32,
}

// this is to prevent a warning that g is not used
impl X {
  pub fn value(self) -> u32 { self.g }
}
```

```rust
// in file b.rs
#![crate_type = "bin"]

extern crate a;

fn main() {
  let x = a::X {};
  x.value();
}
```

Using ` rustc a.rs` followed by `rustc -L . b.rs`, the current output is:

```
error[E0063]: missing fields `f` and `g` in initializer of `X`
 --> b.rs:6:11
  |
6 |   let x = a::X {};
  |           ^^^^ missing `f` and `g`

```

Ideally the output should look like:

```
error: cannot construct `X` with struct literal syntax due to inaccessible fields
 --> b.rs:6:11
  |
6 |   let x = a::X {};
  |           ^^^^

```

The rationale is the same as for https://github.com/rust-lang/rust/issues/76077. Following the suggestion provided by the compiler will lead to another error. This can be frustrating with structs that have a large number of fields or fields with complex types, since the effort of adding those fields to the struct literal is in vain.

By removing the field *f* from the struct, the compiler is reporting that *X* cannot be constructed, as expected.

Tested on both `rustc 1.56.0-nightly (574d37568 2021-08-07)` and `rustc 1.52.0-nightly (4a8b6f708 2021-03-11)`.

**Comment by TheWastl** [comment]
I guess this also applies to patterns:

```rust
pub mod foo {
    #[derive(Default)]
    pub struct Foo {
        pub x: i32,
        y: i32,
    }
}

fn main() {
    let foo::Foo {} = foo::Foo::default();
}
```
[Playground link](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=8ac3a737ae3434f7e5d6d4693049ef08)

Error:

```
error[E0027]: pattern does not mention fields `x`, `y`
  --> src/main.rs:10:9
   |
10 |     let foo::Foo {} = foo::Foo::default();
   |         ^^^^^^^^^^^ missing fields `x`, `y`
   |
help: include the missing fields in the pattern
   |
10 |     let foo::Foo { x, y } = foo::Foo::default();
   |                  ^^^^^^^^
help: if you don't care about these missing fields, you can explicitly ignore them
   |
10 |     let foo::Foo { .. } = foo::Foo::default();
   |                  ^^^^^^

For more information about this error, try `rustc --explain E0027`.
```

Possibly change to something along these lines:


```
error[E0027]: pattern does not mention field `x` and inaccessible fields
  --> src/main.rs:10:9
   |
10 |     let foo::Foo {} = foo::Foo::default();
   |         ^^^^^^^^^^^ missing field `x` and inaccessible fields
   |
help: include the missing fields and ignore the inaccessible fields in the pattern
   |
10 |     let foo::Foo { x, .. } = foo::Foo::default();
   |                  ^^^^^^^^^
help: if you don't care about these missing fields, you can explicitly ignore them
   |
10 |     let foo::Foo { .. } = foo::Foo::default();
   |                  ^^^^^^

For more information about this error, try `rustc --explain E0027`.
```
